### ISSUE TITLE
Constraints Comparing Enum Values

### ISSUE THREAD

**Issue by chapmanjw** [json-schema-org/json-schema-spec#888]
**TL;DR** Is there a simpler way to express that the value of one enum has to be greater or equal to the value of another when the values are non-numeric?

One particular use-case I am looking to solve here is to constraint the values of one enum based on the values of another. In this example, using enums to express a range of non-numeric sizes, looking to ensure the "to" size is greater or equal to the "from" size.  So far, the only way I have found to express this is by permitting all possible enum values of the "to" size and applying the applicable list of enum values on the "from" size.

```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "additionalProperties": false,
  "required": [
    "size_range"
  ],
  "properties": {
    "size_range": {
      "title": "Size Range",
      "type": "object",
      "required": [
        "from"
      ],
      "properties": {
        "from": {
          "title": "From",
          "enum": [
            "XX-Small",
            "X-Small",
            "Small",
            "Medium",
            "Large",
            "X-Large",
            "XX-Large"
          ],
          "type": "string"
        },
        "to": {
          "title": "To",
          "enum": [
            "XX-Small",
            "X-Small",
            "Small",
            "Medium",
            "Large",
            "X-Large",
            "XX-Large"
          ],
          "type": "string"
        }
      },
      "allOf": [
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "XX-Small"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "XX-Small",
                  "X-Small",
                  "Small",
                  "Medium",
                  "Large",
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "X-Small"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "X-Small",
                  "Small",
                  "Medium",
                  "Large",
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "Small"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "Small",
                  "Medium",
                  "Large",
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "Medium"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "Medium",
                  "Large",
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "Large"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "Large",
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "X-Large"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "X-Large",
                  "XX-Large"
                ]
              }
            }
          }
        },
        {
          "if": {
            "properties": {
              "from": {
                "enum": [
                  "XX-Large"
                ]
              }
            }
          },
          "then": {
            "properties": {
              "to": {
                "enum": [
                  "XX-Large"
                ]
              }
            }
          }
        }
      ]
    }
  }
}
```

Is there a simpler way to do this?

**Comment by awwright** [comment]
Even if the values were numeric, this is considered "consistency validation" or "business logic" and is expected to be implemented by the application. I would not suggest trying to validate this with JSON Schema—if someone enters some nonsensical size range, that's not much different than, say, entering ZIP code mismatching the city: the range of values to check is just larger.

If you have a low number of values, then something like this will get you a validation result you're looking for, sure; but semantically speaking, you're creating different classes of objects depending on the lower-bound-size that's selected.

For some additional information see https://github.com/json-schema-org/json-schema-spec/wiki/Scope-of-JSON-Schema-Validation

**Comment by karenetheridge** [comment]
I think I would do something like this:

```
---
  definitions:
    min_size_x_small:
      enum:
        - x_small
        - small
        - medium
        - large
        - x_large
    min_size_small:
        - small
        - medium
        - large
        - x_large
    min_size_medium:
        - medium
        - large
        - x_large
    min_size_large:
        - large
        - x_large
    max_size_small:
        - small
        - medium
        - large
        - x_large
    max_size_medium:
        - medium
        - large
        - x_large
    max_size_large:
        - large
        - x_large
    max_size_x_large:
        - x_large
```

Then you could do something like:

```
  allOf:
    - $ref: #/definitions/min_size_large
    - $ref: #/definitions/max_size_x_large
```

**Comment by Relequestual** [comment]
Sorry, but there's no "cleaner" way to do this with JSON Schema. It's not really a consideration.
Given this issue isn't going to result in any action from us, I'm going to close it, but feel free to join our JSON Schema slack server, and chat and discuss this further should you wish.

**Comment by chapmanjw** [comment]
Thanks for confirming @Relequestual
